File: mini-cpu/PROG/COMPILER/Tiny_Compilo.ml
open Str

let zeros = "00000000"

let const_def = regexp "@\\([a-zA-Z0-9_]+\\)[ ]*=[ ]*\\([0-9]+\\|0x[0-9A-Fa-f]+\\)"
let label_def = regexp "@\\([a-zA-Z0-9_]+\\):"
let sta_instr = regexp "STA *@\\([a-zA-Z0-9_]+\\)"
let jcc_instr = regexp "JCC(@\\([a-zA-Z0-9_]+\\))"
let jmp_instr = regexp "JMP(@\\([a-zA-Z0-9_]+\\))"
let jce_instr = regexp "JCE(@\\([a-zA-Z0-9_]+\\))"
let jcn_instr = regexp "JCN(@\\([a-zA-Z0-9_]+\\))"

let alu_instr = regexp "ACCU *= *ACCU *\\(NOR\\|ADD\\|SUB\\|AND\\|OR\\|XOR\\) *mem\\[@\\([a-zA-Z0-9_]+\\)\\]"
let unary_instr = regexp "ACCU *= *\\(NOT\\|CLR\\) *ACCU"
let mov_instr = regexp "ACCU *= *mem\\[@\\([a-zA-Z0-9_]+\\)\\]"

let mem_reg = Hashtbl.create 32

let to_bin n =
  let rec aux n acc =
    if n = 0 then acc else aux (n lsr 1) (string_of_int (n land 1) ^ acc)
  in
  let s = if n = 0 then "0" else aux n "" in
  String.make (6 - String.length s) '0' ^ s

let read_file filename =
  let ic = open_in filename in
  let rec loop pc acc =
    try
      let line = input_line ic in
      let line = Str.replace_first (regexp ";.*$") "" line |> String.trim in

      if line = "" then loop pc acc

      else if string_match const_def line 0 then
        let name = matched_group 1 line in
        let v = matched_group 2 line in
        let value = int_of_string v in
        Hashtbl.add mem_reg name (to_bin value);
        loop pc acc

      else if string_match label_def line 0 then
        let name = matched_group 1 line in
        Hashtbl.add mem_reg name (to_bin pc);
        loop pc acc

      else
        loop (pc+1) (acc @ [line])

    with End_of_file ->
      close_in ic;
      acc
  in
  loop 0 []

let compile instrs =
  List.iter (fun instr ->

    if string_match sta_instr instr 0 then
      print_endline ("1000" ^ zeros ^ Hashtbl.find mem_reg (matched_group 1 instr))

    else if string_match jmp_instr instr 0 then
      print_endline ("1001" ^ zeros ^ Hashtbl.find mem_reg (matched_group 1 instr))

    else if string_match jcc_instr instr 0 then
      print_endline ("1100" ^ zeros ^ Hashtbl.find mem_reg (matched_group 1 instr))

    else if string_match jce_instr instr 0 then
      print_endline ("1010" ^ zeros ^ Hashtbl.find mem_reg (matched_group 1 instr))

    else if string_match jcn_instr instr 0 then
      print_endline ("1011" ^ zeros ^ Hashtbl.find mem_reg (matched_group 1 instr))

    else if string_match alu_instr instr 0 then
      let op = matched_group 1 instr in
      let addr = Hashtbl.find mem_reg (matched_group 2 instr) in
      let codeop =
        if op = "NOR" then "0000"
        else if op = "ADD" then "0100"
        else if op = "SUB" then "0101"
        else if op = "AND" then "0110"
        else if op = "OR" then "0111"
        else if op = "XOR" then "1000"
        else "0000"
      in
      print_endline (codeop ^ zeros ^ addr)

    else if string_match unary_instr instr 0 then
      let op = matched_group 1 instr in
      let codeop = if op = "NOT" then "0011" else "1111" in
      print_endline (codeop ^ zeros ^ "000000")

    else if string_match mov_instr instr 0 then
      print_endline ("1010" ^ zeros ^ Hashtbl.find mem_reg (matched_group 1 instr))

    else
      print_endline "ERROR"

  ) instrs

let () =
  let instrs = read_file "asm.txt" in
  compile instrs

File: mini-cpu/PROG/COMPILER/Tiny_Compilo.cpp
#include <bits/stdc++.h>
using namespace std;

string zeros = "00000000";

int main() {

    regex const_def(R"(@(\w+)\s*=\s*(\d+|0x[0-9A-Fa-f]+))");
    regex label_def(R"(@(\w+):)");
    regex sta_instr(R"(\s*STA\s+@(\w+))");
    regex jcc_instr(R"(\s*JCC\(@(\w+)\))");
    regex jmp_instr(R"(\s*JMP\(@(\w+)\))");
    regex jce_instr(R"(\s*JCE\(@(\w+)\))");
    regex jcn_instr(R"(\s*JCN\(@(\w+)\))");

    regex alu_instr(R"(\s*ACCU\s*=\s*ACCU\s+(NOR|ADD|SUB|AND|OR|XOR)\s+mem\[@(\w+)\])");
    regex unary_instr(R"(\s*ACCU\s*=\s*(NOT|CLR)\s+ACCU)");
    regex mov_instr(R"(\s*ACCU\s*=\s*mem\[@(\w+)\])");

    map<string, string> mem_reg;
    int PC = 0;

    ifstream file("asm.txt");
    vector<string> instr_set;
    string instr;
    smatch match;

    while (getline(file, instr)) {
        instr = regex_replace(instr, regex(";.*$"), "");
        instr = regex_replace(instr, regex("^\\s+|\\s+$"), "");

        if (instr.empty())
            continue;

        else if (regex_match(instr, match, const_def)) {
            string name = match[1].str();
            string val_str = match[2].str();
            int value = (val_str.substr(0,2) == "0x") ? stoi(val_str, nullptr, 16) : stoi(val_str);
            mem_reg[name] = bitset<6>(value).to_string();
        }

        else if (regex_match(instr, match, label_def)) {
            string name = match[1].str();
            mem_reg[name] = bitset<6>(PC).to_string();
        }

        else {
            instr_set.push_back(instr);
            PC++;
        }
    }

    PC = 0;

    while (PC < instr_set.size()) {
        instr = instr_set[PC];

        if (regex_match(instr, match, sta_instr)) {
            cout << "1000" << zeros << mem_reg[match[1].str()] << endl;
        }

        else if (regex_match(instr, match, jmp_instr)) {
            cout << "1001" << zeros << mem_reg[match[1].str()] << endl;
        }

        else if (regex_match(instr, match, jcc_instr)) {
            cout << "1100" << zeros << mem_reg[match[1].str()] << endl;
        }

        else if (regex_match(instr, match, jce_instr)) {
            cout << "1010" << zeros << mem_reg[match[1].str()] << endl;
        }

        else if (regex_match(instr, match, jcn_instr)) {
            cout << "1011" << zeros << mem_reg[match[1].str()] << endl;
        }

        else if (regex_match(instr, match, alu_instr)) {
            string op = match[1].str();
            string addr = mem_reg[match[2].str()];
            string codeop;

            if (op == "NOR") codeop = "0000";
            else if (op == "ADD") codeop = "0100";
            else if (op == "SUB") codeop = "0101";
            else if (op == "AND") codeop = "0110";
            else if (op == "OR") codeop = "0111";
            else if (op == "XOR") codeop = "1000";

            cout << codeop << zeros << addr << endl;
        }

        else if (regex_match(instr, match, unary_instr)) {
            string op = match[1].str();
            string codeop = (op == "NOT") ? "0011" : "1111";
            cout << codeop << zeros << "000000" << endl;
        }

        else if (regex_match(instr, match, mov_instr)) {
            cout << "1010" << zeros << mem_reg[match[1].str()] << endl;
        }

        else {
            cout << "ERROR" << endl;
        }

        PC++;
    }

    return 0;
}